### DISCUSSION TITLE
Some task stays in queued/None state forever

### DISCUSSION THREAD

**Discussion by vivek-zeta** [apache/airflow#21861] [Q&A]
### Apache Airflow version

2.2.2

### What happened

Airflow version : 2.2.2
I'm using Kubernates executor to run tasks. 
Issue : 
- Some tasks stays in queued/None state forever. 
- While other tasks in same DAG-run not facing this issue.
- It's not happening with specific tasks every time. 
- It is happening with different tasks with different DAG-runs.
- Scheduler logs says task putted in queue. After that no update on that task.
- Seems issue in scheduler.
- Logs for task check_subscription_sync_status which gone in queued state for forever.

Feb 22, 2022 @ 16:05:29.733 | [2022-02-22 10:35:29,732] {scheduler_job.py:538} INFO - Setting external_id for <TaskInstance: orchestra-entitytracker.check_subscription_sync_status job-7012c623-552d-4988-8526-2781f7c66259-ledger-5969073403237049211 [queued]> to 11061
-- | --

 Feb 22, 2022 @ 16:05:29.723 | [2022-02-22 10:35:29,722] {scheduler_job.py:510} INFO - Executor reports execution of orchestra-entitytracker.check_subscription_sync_status run_id=job-7012c623-552d-4988-8526-2781f7c66259-ledger-5969073403237049211 exited with status queued for try_number 1
-- | --

 Feb 22, 2022 @ 16:05:29.420 | [2022-02-22 10:35:29,420] {kubernetes_executor.py:297} INFO - Kubernetes job is (TaskInstanceKey(dag_id='orchestra-entitytracker', task_id='check_subscription_sync_status', run_id='job-7012c623-552d-4988-8526-2781f7c66259-ledger-5969073403237049211', try_number=1), ['airflow', 'tasks', 'run', 'orchestra-entitytracker', 'check_subscription_sync_status', 'job-7012c623-552d-4988-8526-2781f7c66259-ledger-5969073403237049211', '--local', '--subdir', 'DAGS_FOLDER/dags/orchestra-entity.py'], None, None)
-- | --

  Feb 22, 2022 @ 16:05:29.419 | [2022-02-22 10:35:29,419] {kubernetes_executor.py:531} INFO - Add task TaskInstanceKey(dag_id='orchestra-entitytracker', task_id='check_subscription_sync_status', run_id='job-7012c623-552d-4988-8526-2781f7c66259-ledger-5969073403237049211', try_number=1) with command ['airflow', 'tasks', 'run', 'orchestra-entitytracker', 'check_subscription_sync_status', 'job-7012c623-552d-4988-8526-2781f7c66259-ledger-5969073403237049211', '--local', '--subdir', 'DAGS_FOLDER/dags/orchestra-entity.py'] with executor_config {}
-- | --

  Feb 22, 2022 @ 16:05:29.417 | [2022-02-22 10:35:29,416] {base_executor.py:82} INFO - Adding to queue: ['airflow', 'tasks', 'run', 'orchestra-entitytracker', 'check_subscription_sync_status', 'job-7012c623-552d-4988-8526-2781f7c66259-ledger-5969073403237049211', '--local', '--subdir', 'DAGS_FOLDER/dags/orchestra-entity.py']
-- | --

  Feb 22, 2022 @ 16:05:29.417 | [2022-02-22 10:35:29,416] {scheduler_job.py:450} INFO - Sending TaskInstanceKey(dag_id='orchestra-entitytracker', task_id='check_subscription_sync_status', run_id='job-7012c623-552d-4988-8526-2781f7c66259-ledger-5969073403237049211', try_number=1) to executor with priority 6 and queue default
-- | --

  Feb 22, 2022 @ 16:05:29.414 | <TaskInstance: orchestra-entitytracker.check_subscription_sync_status job-7012c623-552d-4988-8526-2781f7c66259-ledger-5969073403237049211 [scheduled]>
-- | --

  Feb 22, 2022 @ 16:05:29.410 | <TaskInstance: orchestra-entitytracker.check_subscription_sync_status job-7012c623-552d-4988-8526-2781f7c66259-ledger-5969073403237049211 [scheduled]>
-- | --

### What you expected to happen

Tasks must not stay in queued state for long time.

### Operating System

k8s

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==2.4.0
apache-airflow-providers-microsoft-azure==2.0.0
apache-airflow-providers-postgres==2.4.0
apache-airflow-providers-docker==2.2.0

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by vivek-zeta** [comment]
@eladkal any thing is known issue like this before?

**Comment by vumdao** [comment]
I got same issue on my airflow cluster version 2.2.1. Any help or advice to troubleshoot appreciated

There're frequent job task stuck in queued state forever although I check the task is in airflow database
```
airflow=> select * from task_instance where task_id='Region_Job_Tasks.1001_Region' and run_id='manual__2022-03-30T05:42:25.399047+00:00';
                 task_id                 |                 dag_id                  |                  run_id                  | start_date | end_date | duration | state  | try_number | hostname | unixname | job_
id |      pool       |  queue  | priority_weight |    operator    |          queued_dttm          | pid | max_tries | executor_config | pool_slots | queued_by_job_id |         external_executor_id         | trig
ger_id | trigger_timeout | next_method | next_kwargs 
-----------------------------------------+-----------------------------------------+------------------------------------------+------------+----------+----------+--------+------------+----------+----------+-----
---+-----------------+---------+-----------------+----------------+-------------------------------+-----+-----------+-----------------+------------+------------------+--------------------------------------+-----
-------+-----------------+-------------+-------------
 Region_Job_Tasks.1001_Region | dev.cloudopz.co_run_all_reports | manual__2022-03-30T05:42:25.399047+00:00 |            |          |          | queued |          0 |          | airflow  |     
   | run-all-reports | default |               2 | PythonOperator | 2022-03-30 05:42:54.130738+00 |     |         1 | \x80057d942e    |          1 |          2005367 | b2e644a7-35d5-44b1-a5a1-3ce8359cb200 |     
       |                 |             | null
(1 row)
```

![image](https://user-images.githubusercontent.com/37215642/161091789-ce3aff6f-6cde-4733-8e08-31587df440f9.png)

Note: restarting schedulers fixed the issue

**Reply by dima-asana** [reply]
Experiencing same symptom with CeleryExecutor.  Airflow 2.1.4, Celery 4.4.7, about 1 task per 200k gets stuck in queued state forever and can't figure out what's causing it.   Also consistent with https://github.com/apache/airflow/issues/19699

For us some (but not all) the queued tasks can be found in the celery reserved queue forever:
```
>>> from airflow.configuration import conf
>>> celery_configuration = conf.getimport('celery', 'celery_config_options')
>>> from celery import Celery, Task, states as celery_states
>>> app = Celery(conf.get('celery', 'CELERY_APP_NAME'), config_source=celery_configuration)
>>> i = app.control.inspect()
>>> i.reserved()
{'celery@worker-74854966d9-q9wgl': [], 'celery@worker-74854966d9-46skn': [], 'celery@worker-74854966d9-tzw7t': [], 'celery@worker-74854966d9-4l7c9': [{'id': '90ae4f62-5064-4597-ac4c-cd43a7454fb8', 'name': 'airflow.executors.celery_executor.execute_command', 'args': [['airflow', 'tasks', 'run', 'qmr_sampling_daily', 'wait_for_emr_redshift_loads-SessionsProcessor', '2022-04-04T00:00:00+00:00', '--local', '--pool', 'sensor', '--subdir', '/var/asana-config/asana/asana2/data/pipelines/airflow/qmr_sampling_daily_dag.py']], 'kwargs': {}, 'type': 'airflow.executors.celery_executor.execute_command', 'hostname': 'celery@worker-74854966d9-4l7c9', 'time_start': None, 'acknowledged': False, 'delivery_info': {'exchange': '', 'routing_key': 'default', 'priority': None, 'redelivered': False}, 'worker_pid': None}], 'celery@worker-74854966d9-tqlh9': [], 'celery@worker-74854966d9-wchpj': []}
```

Will try bringing back automated scheduler restarts, thanks for the suggestion.

Edit: confirmed restarting scheduler fixes the issue for us as well, though it takes ~5 minutes after the restart to fix.  May have been fixed due to the [orphaned task check](https://airflow.apache.org/docs/apache-airflow/2.1.4/configurations-ref.html?highlight=configuration#orphaned-tasks-check-interval).

**Reply by potiuk** [reply]
Can you please upgrade to latest released Airflow - there were a number of fixes since regarding queues and stuck tasks. Hard to pinpoint them exactly, but in any case you will have to upgrade to get any fixes applied (because fixes are implemented in the new versions).

**Reply by dima-asana** [reply]
> automated scheduler restarts

unfortunately, while running scheduler with 5 loops at a time (e.g. `scheduler -n 5`) fixed this issue, it introduced a more difficult one for us (some tasks get scheduled on multiple workers), so we've reverted that.

> upgrading

I've heard this sentiment a few times around the airflow project, so I wanted to give some context on why we can't do this for every issue.  I recognize your difficult position in trying to manage a project with many moving pieces, and impracticality of knowing whether some issue without a reproducible case is still present or not.

We upgraded to Airflow 2.1.4 from 1.14 in December-February this year (it took 2 months of engineering work).  It will likely take additional ~2 weeks of effort to upgrade up to 2.2.5 right now (in part because the [database migration required for the upgrade](https://github.com/apache/airflow/blob/main/UPDATING.md#airflow-220) is untenable at our task instance scale).  We'll definitely upgrade at some point, but it's difficult to justify the work without knowing it would help.

As far as I can tell the attempted fix for this was [reverted](https://github.com/apache/airflow/pull/21556) and no known fix for https://github.com/apache/airflow/issues/19699 is implemented.  Please let me know if you know of some other fix that is potentially likely to apply here (something like #21316 could be possible, I'll look into that)).  That said, I also recognize that you're a volunteer like everyone else and if your position is "we can't spend our time looking into issues that haven't been reproduced on the latest patch version" (there are several reports on 2.2.x) I totally understand.

I also recognize that Airflow is an open source project and it's our shared responsibility to find a better reproducible case for this if not work on a fix after that.

**Reply by potiuk** [reply]
So - my context then and let me guide on where and how you can help @dima-asana . You seem to understand how open-source works so you should know that contribution back and spending time on properly reporting issue and providing enough details to be able to diagnose the problem is CRUCIAL.

I am not sure where you take the ~90% certainty - to be honest. If maintainers are not able to determine with certainty (because they lack information) that your problem has the same root cause as other's problem, I wonder where your certainty comes from.

In fact we are pretty good at determining widespread problems (because several people experience that and repoort it to us with enough details to know it).  But those problems (yes those are are likely different problems) - only a handful of people experiences them otherwise we would be flooded with similar issues. We are not, This likely means that you have some special circumstances - likely deployment-specific issues or specific DAGs or size, or database or possibly a million other things. And without knowing what it is we are in dark.

Saying "i also have the same issue" without providing - YOUR details, YOUR detailed logs, YOUR description of the circimstances and YOUR analysis and attempts that you've done is just not helpful. Yes you likely spent few minutes on writing a paragraph of text and providing your own judgment that this is "the same problem". But that's not  nearly enough.

So if you cannot upgrade - let me propose different thing. Again and again I encourage people to do so.
if you really want to help - spent 3-4 hours gathering your evidences. Filter out the things that are irrelevant, provide logs and description of your cases, likely correlating logs from different sources and looking at your deployment and DAGs to figure out what is likely triggering it. Tell us what's so special about your system (even if those are just intelligent guesses). Unlike you - we have no access to your system, so (since the problem is likely connected to your system's deplyoyment or specific way of developing DAGs). You are the only one who can know that. We can just guess. Open a new issue with "Likely telated to issue x and y" information. Provide Gists with the information that let us help you.

Rather than saying "I know it's the same issue" - show us your data (filtered and preprared so that we do not have to do your job) and let us assess it. Otherwise it's just not helpful,

BTW. There is this really good statement "if something is painful (upgrade) - do it more often" . As counter-intuitive as it seems - the smaller and more often upgrades you do, the less painful it is overall. One of the resons for your problems was really that you jumped from 1.14 to 2.1.4. If you did it in smaller increments, it would have been way less problematic, So it's quite a bit your fault that you waited that long. I heartily recommend the approach of "upgrade often" to everyone.

If in 6 months we will find a critical security zero-day issue like Log4Shel and fix it in 2.5.0 and you will be in 2.1.0, you will have to upgrade from 2.1.0 to 2.5.0  in a matter of hours - then good luck. But if you will be at 2.4.5 - this will be far less painful. And again the problem will be that you are not doing upgrades frequently enough and not because you are doing it too frequently. 

That's why my first advice to anyone is "upgrade" if you are on pretty old version.

**Reply by potiuk** [reply]
And one more thing @dima-asana the main reason why https://github.com/apache/airflow/pull/21556 was reverted, because there was NOONE who experienced the problem willing to test it in their circumstances. And the maintainer who proposed it asked the people several times at least to try. So I would be really careful quoting that in this context.

**Reply by potiuk** [reply]
Actually - since you are so certain it's the same issue @dima-asana.

Maybe you do want to actually take the #21556 and test if the problem is solved ? Actually that would be the BEST way you can help

**Reply by potiuk** [reply]
Another point @dima-asana https://github.com/apache/airflow/issues/13542#issuecomment-1094889441 - similar issue has been reported by one of the users who actually did upgrade to 2.2.5.

Suddenly your 90% looks like, I don't know, maybe 50% or 25% ? Hard to say.

And BTW. Just to explain to casusal readers.  Editing posts and removing statements (without explaining) that someone comments on before refers to is rather poor approach (especially that you can see history of changes in the "edited" dropdown.

**Comment by thesuperzapper** [comment]
The [User-Community Airflow Helm Chart](https://github.com/airflow-helm/charts/tree/main/charts/airflow) added a new kind of scheduler probe in its version `8.6.0` called ["task creation check"](https://github.com/airflow-helm/charts/blob/main/charts/airflow/docs/faq/monitoring/scheduler-liveness-probe.md#scheduler-task-creation-check), which will restart the scheduler if no new tasks have been scheduled in some threshold.

Hopefully, that will catch these kinds of deadlocks (but I am not 100% confident that it will handle all cases).
